### ISSUE TITLE
Play two postional sound simultaneously

### ISSUE THREAD

**Issue by Ted-EE** [kcat/openal-soft#427]
Hello, I met a problem when I code for positonal sound, i.e. when I called `play_do_mi(-1.0, 0.0, -1.0, 2, 1.0, 1.0, 0.0, -1.0, 2, 1.5);` to play the sound, what I expected was that one sound from (-1.0, 0.0, -1.0) and another sound from (1.0, 0.0, -1.0) would come to me at the same time, but actually, it seemed only one sound come from the front.

```
void play_do_mi(float x0, float y0, float z0, float pitch0, float gain0, float x1, float y1, float z1, float pitch1, float gain1) {
    alutInit(NULL, NULL);
    ALCdevice* device = alcGetContextsDevice(alcGetCurrentContext());
    if (alcIsExtensionPresent(device, "ALC_SOFT_HRTF"))
    {
        LPALCRESETDEVICESOFT alcResetDeviceSOFT = (LPALCRESETDEVICESOFT)alcGetProcAddress(device, "alcResetDeviceSOFT");
        const ALCint attributes[3] = {
            ALC_HRTF_SOFT, ALC_TRUE,
            0
        };
        alcResetDeviceSOFT(device, attributes);

        ALCint hrtfenabled = ALC_FALSE;
        alcGetIntegerv(device, ALC_HRTF_SOFT, 1, &hrtfenabled);
        printf("HRTF is%s enabled\n", hrtfenabled ? "" : " not");
    }
    else
        fprintf(stderr, "Device \"%s\" does not support ALC_SOFT_HRTF\n", alcGetString(device, ALC_DEVICE_SPECIFIER));

    //Source position
    ALfloat Source0Pos[] = { x0, y0, z0 };

    //Source velocity
    ALfloat Source0Vel[] = { 0.0, 0.0, 0.0 };

    //Source position
    ALfloat Source1Pos[] = { x1, y1, z1 };

    //Source velocity
    ALfloat Source1Vel[] = { 0.0, 0.0, 0.0 };


    
    ALuint source[2];
    alGenSources(2, source);

    
    ALuint buffer = alutCreateBufferFromFile("do_stretched.wav");
    alSourcei(source[0], AL_BUFFER, buffer);
    alSourcef(source[0], AL_PITCH, pitch0);
    alSourcef(source[0], AL_GAIN, gain0);
    alSourcefv(source[0], AL_POSITION, Source0Pos);
    alSourcefv(source[0], AL_VELOCITY, Source0Vel);
    alSourcei(source[0], AL_LOOPING, AL_FALSE);


    alSourcei(source[1], AL_BUFFER, buffer);
    alSourcef(source[1], AL_PITCH, pitch1);
    alSourcef(source[1], AL_GAIN, gain1);
    alSourcefv(source[1], AL_POSITION, Source1Pos);
    alSourcefv(source[1], AL_VELOCITY, Source1Vel);
    alSourcei(source[1], AL_LOOPING, AL_FALSE);


    alSourcePlay(source[0]);
    alSourcePlay(source[1]);

    ALint state;
    do {
        alGetSourcei(source[0], AL_SOURCE_STATE, &state);
        alGetSourcei(source[1], AL_SOURCE_STATE, &state);
    } while (state == AL_PLAYING);

    alDeleteSources(2, source);
    alDeleteBuffers(1, &buffer);

    alutExit();
}
```

I am not sure if I made some mistakes in my code. I would appreciate if you could point the mistake out for me or give me some suggestions on playing two sounds from different position simultaneously.

Thanks.

**Comment by kcat** [comment]
> what I expected was that one sound from (-1.0, 0.0, -1.0) and another sound from (1.0, 0.0, -1.0) would come to me at the same time, but actually, it seemed only one sound come from the front.

Because you're playing the same sound on the left and right at the same time, which causes a "phantom sound" in the center. Like if you play the same sound on the left speaker and right speaker at the same time, it sounds like it comes from in between the speakers. You need to use different sounds, or play them at different times, to hear them separately.

**Comment by fbirot** [comment]
In order for hearing the same source at the same time at two different locations, you can also have a look at signal decorrelation with all pass filters. This would permit you to have two sounds, that sound the same, but that have small phase changes (they are not exactly the same, which will allow your hear to hear two sounds, instead of one in the middle)

**Comment by Ted-EE** [comment]
Thanks for these quick answers! I will look into the solution.

**Comment by Ted-EE** [comment]
When I used two different sounds, I could listen both them, but I think, the spatiality may be reduced. Is there any way except playing them at different times that also keeps the spatiality?

Thanks!

**Comment by kcat** [comment]
> When I used two different sounds, I could listen both them, but I think, the spatiality may be reduced.

What do you mean by "the spatiality may be reduced"?

**Comment by Ted-EE** [comment]
Thanks for your replying!

That means, when I played two different sounds at the same time, e.g. one piano note do from front left, the other piano note mi from front right, it was not easy to recognize the positions of both them.

However, I think I found an approach that could be acceptable for my project, that was to generate two sine waves instead of loading wav files of piano notes (do, re, mi...), and changed both AL_PITCH to ensure pitch of one sound was larger than the other. Then I could easily recognized both position, this approach was very similar with "play sound at different times, to hear them separately".

Thanks.
